persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedActorOf.scala

```diff
@@ -0,0 +1,161 @@
+      * 
+      * Persistence layer, used to store/recover events and snapshots, provided by [[EventSourcedPersistence]].  
+      *
+      * @param eventSourcedOf actor logic used to recover and handle messages 
+      * @param persistence persistence used for event sourcing
+      * @return instance of [[Actor]]
+      */
+  def actor[F[_]: Async: ToFuture, S, E, C: ClassTag](
+    eventSourcedOf: EventSourcedOf[F, Lifecycle[F, S, E, C]],
+    persistence: EventSourcedPersistence[F]
+  ): Actor = ActorOf[F] { actorCtx =>
+    for {
+      eventSourced    <- eventSourcedOf(actorCtx).toResource
+      recoveryStarted <- eventSourced.value
+
+      snapshotStore <- persistence.snapshotStore[S](eventSourced).toResource
+      eventStore    <- persistence.eventStore[E](eventSourced).toResource
+
+      snapshot <- snapshotStore.latest.toResource
+
+      recovering <- recoveryStarted(
+        snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min),
+        snapshot.map(_.asOffer)
+      )
+
+      seqNr <- recovering.replay.use { replay =>
+        val snapSeqNr = snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min)
+        val fromSeqNr = snapshot.map(_.metadata.seqNr + 1).getOrElse(SeqNr.Min)
+        for {
+          events <- eventStore.events(fromSeqNr)
+          seqNrL <- events.foldWhileM(snapSeqNr) {
+            case (_, EventStore.Event(event, seqNr)) => replay(event, seqNr).as(seqNr.asLeft[Unit])
+            case (_, EventStore.HighestSeqNr(seqNr)) => seqNr.asLeft[Unit].pure[F]
+          }
+          seqNr <- seqNrL match {
+            case Left(seqNr) => seqNr.pure[F]
+            case Right(_)    =>
+              // function used in events.foldWhileM always returns Left
+              // and sstream.Stream.foldWhileM returns Right only if passed function do so
+              // thus makes getting Right here impossible
+              new IllegalStateException("should never happened").raiseError[F, SeqNr]
+          }
+        } yield seqNr
+      }.toResource
+
+      currentSeqNr <- Ref[F].of(seqNr).toResource
+
+      snapshotter = new Snapshotter[F, S] {
+
+        def save(seqNr: SeqNr, snapshot: S): F[F[Instant]] = snapshotStore.save(seqNr, snapshot)
+
+        def delete(seqNr: SeqNr): F[F[Unit]] = snapshotStore.delete(seqNr)
+
+        def delete(criteria: SnapshotSelectionCriteria): F[F[Unit]] = snapshotStore.delete(criteria.asStoreCriteria)
+
+      }
+
+      journaller = new Journaller[F, E] {
+
+        val append = new Append[F, E] {
+
+          def apply(events: Events[E]): F[F[SeqNr]] =
+            currentSeqNr
+              .modify { seqNr0 =>
+                val seqNr1 = seqNr0 + events.size
+                seqNr1 -> seqNr0
+              }
+              .map { seqNr0 =>
+                var seqNr = seqNr0
+                def nextSeqNr: SeqNr = {
+                  seqNr = seqNr + 1
+                  seqNr
+                }
+                events.map(e => EventStore.Event(e, nextSeqNr))
```

Comment:
Is it worth to add `List#mapAccumulate` equivalent to `Events` class and then reuse it here, to avoid using `var` + `map` combination, or that would be an overkill?

I mean something like following (I did not test it):
```
  def mapAccumulate[S, B](init: S)(f: (S, A) => (S, B)): (S, Events[B]) = {
    val (s, values) =
      this.values.mapAccumulate(init) { (s, a) =>
        a.mapAccumulate(s)(f)
      }
    (s, Events(values))
  }
```

Reply:
`Events` already has instance of `Traverse` thus method `.mapAccumulate` available out of the box, the change was simple & strait forward. Please review one more time